### PR TITLE
Migrate home screen to Mobius - Migrate

### PR DESCRIPTION
**Pull request by msasikanth**
https://app.clubhouse.io/simpledotorg/story/726/migrate-home-screen-to-mobius

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/home/HomeScreenEffectHandler.kt [hunk 1]
```diff
@@ -19,5 +23,16 @@ class HomeScreenEffectHandler @AssistedInject constructor(
...
       .addAction(OpenFacilitySelection::class.java, uiActions::openFacilitySelection, schedulersProvider.ui())
+      .addTransformer(LoadCurrentFacility::class.java, loadCurrentFacility())
       .build()
+
+  private fun loadCurrentFacility(): ObservableTransformer<LoadCurrentFacility, HomeScreenEvent> {
+    return ObservableTransformer { effects ->
+      effects
+          .observeOn(schedulersProvider.io())
+          .flatMap { userSession.requireLoggedInUser() }
+          .flatMap { facilityRepository.currentFacility(it) }
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/home/HomeScreenEffectHandler.kt:52]: In a recent PR (#1598), we added a way to inject a stream of the current facility as `Observable<Facility>`. In the final cleanup of this screen, we might want to switch to using this.

File: app/src/main/java/org/simple/clinic/home/HomeScreenEffectHandler.kt [hunk 2]
```diff
@@ -24,8 +29,21 @@ class HomeScreenEffectHandler @AssistedInject constructor(
...
       .addAction(OpenFacilitySelection::class.java, uiActions::openFacilitySelection, schedulersProvider.ui())
       .addTransformer(LoadCurrentFacility::class.java, loadCurrentFacility())
+      .addTransformer(LoadOverdueAppointmentCount::class.java, loadOverdueAppointmentCount())
       .build()
 
+  private fun loadOverdueAppointmentCount(): ObservableTransformer<LoadOverdueAppointmentCount, HomeScreenEvent> {
+    return ObservableTransformer { effects ->
+      effects
+          .observeOn(schedulersProvider.io())
+          .flatMap { (facility) ->
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/home/HomeScreenEffectHandler.kt:39]: We should ideally be using `.switchMap` by default and only using `.flatMap` where it is required.

File: app/src/main/java/org/simple/clinic/home/HomeScreenUiRenderer.kt [hunk 3]
```diff
@@ -8,6 +8,18 @@ class HomeScreenUiRenderer(private val ui: HomeScreenUi) : ViewRenderer<HomeScre
...
     }
+
+    if (model.hasOverdueAppointmentCount) {
+      renderOverdueAppointmentCount(model)
+    }
+  }
+
+  private fun renderOverdueAppointmentCount(model: HomeScreenModel) {
+    val overdueAppointmentCount = model.overdueAppointmentCount!!
+    if (overdueAppointmentCount > 0)
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/home/HomeScreenUiRenderer.kt:19]: What do you think about making this check a part of `HomeScreenModel` as a computed property, say, `hasAtLeastOneOverdueAppointment`?

File: app/src/main/java/org/simple/clinic/home/HomeScreenEffectHandler.kt (hunk 2 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/home/HomeScreenEffectHandler.kt:39]: Correct me if I am wrong, we need to use `switchMap` to get the latest observable stream right? Does that mean `flatMap` will retain older streams?

File: app/src/main/java/org/simple/clinic/home/HomeScreenUiRenderer.kt (hunk 3 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/home/HomeScreenUiRenderer.kt:19]: Oh yeah, I should have done that. Completely missed it, will do it in final PR 👍🏾

File: app/src/main/java/org/simple/clinic/home/HomeScreenEffectHandler.kt (hunk 2 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/home/HomeScreenEffectHandler.kt:39]: Yes, that's exactly what `flatMap` does.

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/home/HomeScreenEffectHandler.kt:39]: If you look at the source code for [`ObservableFlatMap`](https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/operators/observable/ObservableFlatMap.java) and [`ObservableSwitchMap`](https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/operators/observable/ObservableSwitchMap.java), you'll see the different between the two.

**Pull request merged**